pulp_docker/app/tasks/stages/dedupe_save.py

```diff
@@ -0,0 +1,163 @@
+                :class:`~pulpcore.plugin.stages.DeclarativeContent` objects from.
+            out_q (:class:`asyncio.Queue`): The queue to put
+                :class:`~pulpcore.plugin.stages.DeclarativeContent` into.
+        Returns:
+            The coroutine for this stage.
+
+        """
+        while True:
+            dc = await in_q.get()
+            if dc is None:
+                break
+            self.query_and_save_artifacts(dc)
+            await out_q.put(dc)
+        await out_q.put(None)
+
+    def query_and_save_artifacts(self, dc):
+        """
+        Combine duplicated artifacts, save unique artifacts.
+
+        Args:
+            in_q (:class:`asyncio.Queue`): The queue to receive
+                :class:`~pulpcore.plugin.stages.DeclarativeContent` objects from.
+        """
+        for da in dc.d_artifacts:
+            artifact_q = Q()
+            for digest_name in da.artifact.DIGEST_FIELDS:
+                digest_value = getattr(da.artifact, digest_name)
+                if digest_value:
+                    key = {digest_name: digest_value}
+                    artifact_q &= Q(**key)
+            try:
+                existing_artifact = Artifact.objects.get(artifact_q)
+                # TODO count deduped artifacts
+            except Artifact.DoesNotExist as e:
+                da.artifact.save()
+            else:
+                da.artifact = existing_artifact
+
+
+class SerialContentSave(Stage):
+    """
+    Combine duplicated Content, save unique Content.
+    """
+
+    async def __call__(self, in_q, out_q):
+        """
+        The coroutine for this stage.
+
+        Args:
+            in_q (:class:`asyncio.Queue`): The queue to receive
+                :class:`~pulpcore.plugin.stages.DeclarativeContent` objects from.
+            out_q (:class:`asyncio.Queue`): The queue to put
+                :class:`~pulpcore.plugin.stages.DeclarativeContent` into.
+        Returns:
+            The coroutine for this stage.
+
+        """
+        while True:
+            dc = await in_q.get()
+            # finished
+            if dc is None:
+                break
+            # Artifacts have not been downloaded
+            elif not self.settled(dc):
+                await out_q.put(dc)
+            # already saved
+            elif dc.content.pk is not None:
+                await out_q.put(dc)
+            # needs to be saved
+            else:
+                self.query_and_save_content(dc)
+                await out_q.put(dc)
+        await out_q.put(None)
+
+    def query_and_save_content(self, dc):
+        """
+        Combine duplicate Content, save unique Content.
+
+        Args:
+            dc (class:`~pulpcore.plugin.stages.DeclarativeContent`): Object containing Content to
+                                                                     be saved.
+        """
+        model_type = type(dc.content)
+        unit_key = dc.content.natural_key_dict()
+        try:
```

Comment:
This _query-and-save_ logic is vulnerable to race conditions.  A better design pattern is:
```
try:
    model.save()
except IntegrityError:  # dup
    model = model.objects.get(**natural_key)
```